contractcourt/htlc_timeout_resolver.go

```diff
@@ -733,3 +740,163 @@ func (h *htlcTimeoutResolver) HtlcPoint() wire.OutPoint {
 // A compile time assertion to ensure htlcTimeoutResolver meets the
 // ContractResolver interface.
 var _ htlcContractResolver = (*htlcTimeoutResolver)(nil)
+
+// spendResult is used to hold the result of a spend event from either a
+// mempool spend or a block spend.
+type spendResult struct {
+	// spend contains the details of the spend.
+	spend *chainntnfs.SpendDetail
+
+	// err is the error that occurred during the spend notification.
+	err error
+}
+
+// waitForMempoolSpend waits for the htlc output to be spent by a transaction
+// that's either be found in the mempool or in a block.
+func (h *htlcTimeoutResolver) waitForMempoolSpend(op wire.OutPoint,
+	pkScript []byte) (*chainntnfs.SpendDetail, error) {
+
+	log.Infof("%T(%v): waiting for spent of HTLC output %v to be found "+
+		"in mempool or block", h, h.htlcResolution.ClaimOutpoint, op)
+
+	// Subscribe for block spent(confirmed).
+	blockSpent, err := h.Notifier.RegisterSpendNtfn(
+		&op, pkScript, h.broadcastHeight,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("register spend: %w", err)
+	}
+
+	// Subscribe for mempool spent(unconfirmed).
+	mempoolSpent, err := h.Mempool.SubscribeMempoolSpent(op)
+	if err != nil {
+		return nil, fmt.Errorf("register mempool spend: %w", err)
+	}
+
+	// Create a result chan that will be used to receive the spending
+	// events.
+	result := make(chan *spendResult, 1)
+
+	// Create a goroutine that will wait for either a mempool spend or a
+	// block spend.
+	//
+	// NOTE: no need to use waitgroup here as when the resolver exits, the
+	// goroutine will return on the quit channel.
+	go h.consumeSpendEvents(result, blockSpent.Spend, mempoolSpent.Spend)
+
+	// Wait for the spend event to be received.
+	select {
+	case event := <-result:
+		// Cancel the mempool subscription as we don't need it anymore.
+		mempoolSpent.Cancel()
+
+		return event.spend, event.err
+
+	case <-h.quit:
+		return nil, errResolverShuttingDown
+	}
+}
+
+// consumeSpendEvents consumes the spend events from the block and mempool
+// subscriptions. It exits when a spend event is received from the block, or
+// the resolver itself quits. When a spend event is received from the mempool,
+// however, it won't exit but continuing to wait for a spend event from the
+// block subscription.
+//
+// NOTE: there could be a case where we found the preimage in the mempool,
```

Comment:
I don't think this is an attack because if we know the preimage, any attack is over

Reply:
I think it's an attempted attack but has no effect now(and probably losing money).

Reply:
Unless there is a concrete attack scenario that is either:
- fixed now, or
- still exists
I think we should avoid speculation and remove this comment.

Reply:
Cool changed the docs to describe the current behavior only.

Reply:
Updated the description. Yeah the docs were a bit confusing, now updated. I was trying to describe the possible outcomes and specifically the free money.